From ec93faf228f22afc46c1c9c71762fd4de0943f92 Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@outlook.com>
Date: Mon, 24 Aug 2015 22:37:36 -0700
Subject: [PATCH] Add a flag to disable InventoryMoveItemEvent

Even with no listeners, InventoryMoveItemEvent hurts performance.
This patch prevents the event from firing if there are no listeners
Some plugins offer no way to disable their IMIE listeners. (See WorldGuard-3520)
This patch adds a per-world flag to disable InventoryMoveItemEvent for hoppers.

When the event is disabled, the overhead of cloning items, and various other things are avoided

diff --git a/src/main/java/net/minecraft/server/TileEntityHopper.java b/src/main/java/net/minecraft/server/TileEntityHopper.java
index 8d08ad4..6c0cbe6 100644
--- a/src/main/java/net/minecraft/server/TileEntityHopper.java
+++ b/src/main/java/net/minecraft/server/TileEntityHopper.java
@@ -266,8 +266,13 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
                         ItemStack itemstack = this.getItem(i).cloneItemStack();
                         // ItemStack itemstack1 = addItem(iinventory, this.splitStack(i, 1), enumdirection);
 
+                        // TacoSpigot start - add an option to turn of InventoryMoveItemEvent
+                        final ItemStack originalStack = this.splitStack(i, world.spigotConfig.hopperAmount);
+                        final ItemStack stack;
+                        if (HopperHelper.isFireInventoryMoveItemEvent(this)) {
                         // CraftBukkit start - Call event when pushing items into other inventories
-                        CraftItemStack oitemstack = CraftItemStack.asCraftMirror(this.splitStack(i, world.spigotConfig.hopperAmount)); // Spigot
+                        CraftItemStack oitemstack = CraftItemStack.asCraftMirror(originalStack); // Spigot
+                        // TacoSpigot end
 
                         Inventory destinationInventory;
                         // Have to special case large chests as they work oddly
@@ -284,11 +289,23 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
                             this.d(world.spigotConfig.hopperTransfer); // Spigot
                             return false;
                         }
-                        int origCount = event.getItem().getAmount(); // Spigot
-                        ItemStack itemstack1 = addItem(iinventory, CraftItemStack.asNMSCopy(event.getItem()), enumdirection);
+                        // TacoSpigot start
+                        stack = CraftItemStack.asNMSCopy(event.getItem());
+                        // handle cases where the event is not fired
+                        } else {
+                            /*
+                             * Normally, vanilla would require us to call 'iinventory.update()'
+                             * craftbukkit logic below calls update() if 'stack.equals(originalStack)'
+                             * Since we set 'stack' to be the same as 'orgiginalStack' they will be eaual, and update
+                             */
+                            stack = originalStack;
+                        }
+                        int origCount = stack.count; // Spigot
+                        ItemStack itemstack1 = addItem(iinventory, stack, enumdirection);
+                        // TacoSpigot end
 
                         if (itemstack1 == null || itemstack1.count == 0) {
-                            if (event.getItem().equals(oitemstack)) {
+                            if (stack.equals(originalStack)) { // TacoSpigot
                                 iinventory.update();
                             } else {
                                 this.setItem(i, itemstack);
@@ -417,8 +434,13 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
         if (itemstack != null && b(iinventory, itemstack, i, enumdirection)) {
             ItemStack itemstack1 = itemstack.cloneItemStack();
             // ItemStack itemstack2 = addItem(ihopper, iinventory.splitStack(i, 1), (EnumDirection) null);
+            // TacoSpigot start - add an option to turn of InventoryMoveItemEvent
+            final ItemStack originalStack = iinventory.splitStack(i, ihopper.getWorld().spigotConfig.hopperAmount);
+            final ItemStack stack;
+            if (HopperHelper.isFireInventoryMoveItemEvent(ihopper)) {
             // CraftBukkit start - Call event on collection of items from inventories into the hopper
-            CraftItemStack oitemstack = CraftItemStack.asCraftMirror(iinventory.splitStack(i, ihopper.getWorld().spigotConfig.hopperAmount)); // Spigot
+            CraftItemStack oitemstack = CraftItemStack.asCraftMirror(originalStack); // Spigot
+            // TacoSpigot end
 
             Inventory sourceInventory;
             // Have to special case large chests as they work oddly
@@ -441,11 +463,23 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
                 }
                 return false;
             }
-            int origCount = event.getItem().getAmount(); // Spigot
-            ItemStack itemstack2 = addItem(ihopper, CraftItemStack.asNMSCopy(event.getItem()), null);
+            // TacoSpigot start
+            stack = CraftItemStack.asNMSCopy(event.getItem());
+            // handle cases where the event is not firedgol
+            } else {
+                /*
+                 * Normally, vanilla would require us to call 'iinventory.update()'
+                 * craftbukkit logic below calls update() if 'stack.equals(originalStack)'
+                 * Since we set 'stack' to be the same as 'orgiginalStack' they will be eaual, and update
+                 */
+                stack = originalStack;
+            }
+            int origCount = stack.count; // Spigot
+            ItemStack itemstack2 = addItem(ihopper, stack, null);
+            // TacoSpigot end
 
             if (itemstack2 == null || itemstack2.count == 0) {
-                if (event.getItem().equals(oitemstack)) {
+                if (stack.equals(originalStack)) { // TacoSpigot
                     iinventory.update();
                 } else {
                     iinventory.setItem(i, itemstack1);
diff --git a/src/main/java/net/techcable/tacospigot/HopperHelper.java b/src/main/java/net/techcable/tacospigot/HopperHelper.java
index 4eb5430..1250bf9 100644
--- a/src/main/java/net/techcable/tacospigot/HopperHelper.java
+++ b/src/main/java/net/techcable/tacospigot/HopperHelper.java
@@ -11,6 +11,7 @@ import net.minecraft.server.ChunkSection;
 import net.minecraft.server.Entity;
 import net.minecraft.server.EntityItem;
 import net.minecraft.server.IBlockData;
+import net.minecraft.server.IHopper;
 import net.minecraft.server.IInventory;
 import net.minecraft.server.InventoryLargeChest;
 import net.minecraft.server.MathHelper;
@@ -21,6 +22,7 @@ import net.minecraft.server.World;
 import net.techcable.tacospigot.utils.BlockHelper;
 
 import org.bukkit.Material;
+import org.bukkit.event.inventory.InventoryMoveItemEvent;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -72,4 +74,8 @@ public class HopperHelper {
         }
         return null;
     }
+
+    public static boolean isFireInventoryMoveItemEvent(IHopper hopper) {
+        return hopper.getWorld().tacoSpigotConfig.isHopperFireIMIE && InventoryMoveItemEvent.getHandlerList().getRegisteredListeners().length > 0;
+    }
 }
diff --git a/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java b/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java
index c012c8d..c1f297c 100644
--- a/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java
+++ b/src/main/java/net/techcable/tacospigot/TacoSpigotWorldConfig.java
@@ -69,6 +69,11 @@ public class TacoSpigotWorldConfig {
         isHopperPushBased = getBoolean("hopper.push-based", true);
     }
 
+    public boolean isHopperFireIMIE;
+    private void isHopperFireIMIE() {
+        isHopperFireIMIE = getBoolean("hopper.fire-InventoryMoveItemEvent", true);
+    }
+
     public boolean optimizeArmorStandMovement;
     private void isArmorStandMoveWithoutGravity() {
         optimizeArmorStandMovement = getBoolean("armor-stand.optimize-movement", false); // Doesn't fully emulate vanilla behavior, see issue #1
-- 
2.6.4

